src/validators/vtrValidator.ts

```diff
@@ -0,0 +1,113 @@
+} from "../constants/providerConfig.js";
+import { Config } from "../config.js";
+import { logger } from "../logger.js";
+
+export const vtrValidator = (
+  ctx: KoaContextWithOIDC,
+  vtr: string | undefined,
+  client: Client
+): void => {
+  const parsedVtr = parseVtr(vtr, ctx);
+  const clientLocs = Config.getInstance().getClientLoCs();
+  const clientIdVerificationEnabled =
+    Config.getInstance().getIdentityVerificationSupported();
+
+  const splitVtrValues = parsedVtr.split(".");
+  const credentialTrustValue = splitVtrValues
+    .filter(isCredentialTrust)
+    .join(".");
+
+  const levelOfConfidence = splitVtrValues.filter(isVectorOfTrust)[0];
+  const invalidValues = splitVtrValues.filter(isInvalid);
+
+  if (invalidValues.length > 0) {
+    logger.error("Invalid Credential trust values present in Vtr", parsedVtr);
+    throw new errors.InvalidRequest("Request vtr not valid", 302);
+  }
+  const isValidCredentialTrust =
+    VALID_CREDENTIAL_TRUST_VALUES.includes(credentialTrustValue);
+  if (!isValidCredentialTrust) {
+    logger.error("Invalid Credential trust values present in Vtr", vtr);
+    throw new errors.InvalidRequest("Request vtr not valid", 302);
+  }
+
+  if (!levelOfConfidence) {
+    //Valid Credential trust and no LOC provided
+    return;
+  }
+
+  const isValidLoc = VALID_LOC_VALUES.includes(levelOfConfidence);
+  const isLocInConfig = clientLocs.includes(levelOfConfidence);
+
+  if (!isValidLoc) {
+    logger.error("Invalid Level of confidence specified", levelOfConfidence);
+    throw new errors.InvalidRequest("Request vtr not valid", 302);
+  }
+
+  if (!isLocInConfig) {
+    logger.error(
+      "Invalid Level of confidence for client",
+      levelOfConfidence,
+      clientLocs.join(", ")
+    );
+    throw new errors.InvalidRequest("Request vtr not valid", 302);
+  }
+
+  if (levelOfConfidence === "P0" && isLocInConfig) {
+    //requested P0 and client configured for P0
+    return;
+  }
+
+  if (
+    levelOfConfidence === "P2" &&
+    credentialTrustValue === "Cl.Cm" &&
+    clientIdVerificationEnabled
+  ) {
+    return;
+  } else {
+    logger.error(
+      "Invalid request",
+      JSON.stringify(
+        {
+          requestVtr: parsedVtr,
+          client_id: client.clientId,
+          isIdVerifyEnabledForClient: clientIdVerificationEnabled,
+        },
+        null,
+        4
+      )
+    );
+    throw new errors.InvalidRequest("Request vtr not valid", 302);
+  }
+};
+
+const parseVtr = (vtr: string | undefined, ctx: KoaContextWithOIDC): string => {
+  if (!vtr) {
+    logger.info("Vtr not present, replacing with default");
+    (ctx.oidc.params as Record<string, unknown>).vtr = JSON.stringify([
+      DEFAULT_CREDENTIAL_TRUST,
+    ]);
+    return DEFAULT_CREDENTIAL_TRUST;
+  }
+  let parsedVtr: string[];
+  try {
+    parsedVtr = JSON.parse(vtr);
+  } catch (error) {
+    logger.error("Error parsing VTR");
+    throw new errors.InvalidRequest("Request vtr not valid", 302);
```

Comment:
The type specified that code should be a number so I assumed it meant a status code, which it sort of looks like it might be here, https://github.com/panva/node-oidc-provider/blob/3d1dfe4eba3bb56e65b3e0f7a4321ec613e3a074/lib/helpers/errors.js#L65 but it doesn't make any difference when testing the endpoint, so will remove